lib/sdoblockio.c

```diff
@@ -142,911 +69,328 @@ void sdo_resize_block(sdo_block_t *sdob, int need)
-	return n >> 1;
+bool sdow_encoded_length(sdow_t *sdow, size_t *length) {
+	*length = cbor_encoder_get_buffer_size(&sdow->current->cbor_encoder, sdow->b.block);
+	return true;
 }
-#endif
 
-/**
- * Reads a byte array base64 into the buffer provided
- */
-int sdo_read_byte_array_field(sdor_t *sdor, int b64Sz, uint8_t *bufp,
-			      int buf_sz)
+void sdow_flush(sdow_t *sdow)
 {
-	int converted = 0;
-
-	if (!_read_comma(sdor)) {
-		LOG(LOG_ERROR, "we were expecting , here!\n");
-		goto err;
-	}
-
-	// LOG(LOG_ERROR, "SDORead_byte_array\n");
-	if (!_read_expected_char(sdor, '"'))
-		goto err;
-
-	converted = b64To_bin((size_t)b64Sz, sdor->b.block, sdor->b.cursor,
-			      (size_t)buf_sz, bufp, 0);
-
-	if (converted == -1) {
-		LOG(LOG_ERROR, "Base64 string is invalid!\n");
-		goto err;
-	}
-	sdor->b.cursor += b64Sz;
-
-	if (!_read_expected_char(sdor, '"'))
-		goto err;
-
-	sdor->need_comma = true;
-
-	return converted;
-
-err:
-	return 0; /* Any failure means no bytes read */
+	sdo_block_t *sdob = &sdow->b;
+	sdo_block_reset(sdob);
 }
 
 //==============================================================================
-// Write values
+// Read values
 //
 
 /**
- * SDOW - SDO Writer
+ * Initialize SDO CBOR packet reader engine
+ *
+ * @param sdor - Pointer of struct containing SDOR data structure,
+ *
+ * @return
+ *        return true on success. false on failure.
  */
-bool sdow_init(sdow_t *sdow)
+bool sdor_init(sdor_t *sdor)
 {
-	if (memset_s(sdow, sizeof(*sdow), 0) != 0) {
-		LOG(LOG_ERROR, "SDOW memset() failed!\n");
+	if (memset_s(sdor, sizeof(*sdor), 0) != 0) {
+		LOG(LOG_ERROR, "SDOR memset() failed!\n");
 		return false;
 	}
-
-	sdo_block_init(&sdow->b);
-
+	sdo_block_reset(&sdor->b);
 	return true;
 }
 
-/**
- * Internal API
- */
-void sdow_block_reset(sdow_t *sdow)
-{
-	sdo_block_t *sdob = &sdow->b;
+bool sdor_parser_init(sdor_t *sdor, sdo_block_t *received_block) {
+	sdor->current = malloc(sizeof(sdor_cbor_decoder_t));
+	sdor->current->next = NULL;
+	sdor->current->previous = NULL;
 
-	sdob->cursor = sdob->block_size = 0;
-	sdow->need_comma = false;
-}
+	memcpy_s(sdor->b.block, CBOR_BUFFER_LENGTH, received_block->block, received_block->block_size);
 
-/**
- * Internal API
- */
-int sdow_next_block(sdow_t *sdow, int type)
-{
-	sdow_block_reset(sdow);
-	sdow->msg_type = type;
+    if (cbor_parser_init(sdor->b.block, sdor->b.block_size, 0, &sdor->cbor_parser,
+	 	&sdor->current->cbor_value) != CborNoError) {
+		LOG(LOG_ERROR, "CBOR decoder: Failed to initialize CBOR Parser\n");
+		return false;
+	}
 	return true;
 }
 
-/**
- * Internal API
- */
-static void _write_comma(sdow_t *sdow)
-{
-	sdo_block_t *sdob = &sdow->b;
-
-	if (sdow->need_comma) {
-		sdow->need_comma = false;
-		sdoBPutC(sdob, ',');
-		if (sdob->block_size < sdob->cursor)
-			sdob->block_size = sdob->cursor;
+bool sdor_start_array(sdor_t *sdor) {
+	// create next, create backlink and move forward.
+	sdor->current->next = sdo_alloc(sizeof(sdor_cbor_decoder_t));
```

Comment:
At some places malloc() is used, whereas at some places sdo_alloc() is used. Switch to one of them for uniformity.